### PR TITLE
refactor: remove suspicious mutation

### PR DESCRIPTION
**Pull request by rgrinberg**

### CODE REVIEW

File: src/dune_engine/build_system.ml [hunk 1]
```diff
@@ -1133,31 +1133,33 @@ let files_of ~dir =
...
-    List.fold_left sets ~init:Package.Id.Set.empty ~f:Package.Id.Set.union
+          List.filter_map ~f:Path.as_in_build_dir)
+  and loop_files rules_seen files =
+    Memo.List.fold_left ~init:(Package.Id.Set.empty, rules_seen) files
+      ~f:(fun (sets, rules_seen) file ->
+        let+ set, rules_seen = loop rules_seen file in
+        (Package.Id.Set.union set sets, rules_seen))
   in
-  loop_files files
+  let+ packages, _rules_seen = loop_files Rule.Set.empty files in
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:1161]: Do we actually need to return `rules_seen`? It seems to me that we only need to accumulate them to pass into recursive calls but never need the returned values.

Comment (snowleopard) [reply src/dune_engine/build_system.ml:1161]: Never mind, it looks like we actually need the returned values too.

Comment (rgrinberg) [reply src/dune_engine/build_system.ml:1161]: We can discard it in the end, but I think we need in the intermediate calls that accumulate the seen rules.

**Review by snowleopard — APPROVED** [review]
Looks good!

**Pull request closed without merging**